floqast_sftp/app.py

```diff
@@ -0,0 +1,273 @@
+    -------
+    datetime.date
+        Date one month previous to the given date.
+
+    """
+    if _date.month == 1:
+        _year = _date.year - 1
+        return _date.replace(year=_year, month=12)
+
+    _month = _date.month - 1
+    return _date.replace(month=_month)
+
+
+def get_balances_csv(when):
+    """
+    Fetch the balances CSV from lambda-mips-api for the given activity period.
+
+    Parameters
+    ----------
+    when: datetime.date
+        Date to pass to lambda-mips-api for calculating the balance activity period.
+
+    Returns
+    -------
+    (str, file)
+        Tuple of file name and a file-like object.
+
+    """
+    url = _csv_url + f"&target_date={when}"
+
+    response = requests.get(url, stream=True)
+    response.raise_for_status()
+    file_obj = io.StringIO(response.text)
+
+    # read target period from CSV
+    csv_dict = csv.DictReader(file_obj)
+    row = next(csv_dict)
+    period = row["PeriodStart"]
+    # and reset file position
+    file_obj.seek(0)
+
+    filename = get_file_name(period)
+
+    return filename, file_obj
+
+
+def put_sftp_file(client, name, file_obj):
+    """
+    Wrapper around SFTPClient.putfo
+
+    Parameters
+    ----------
+    client: paramiko.SFTPClient
+        SFTP client.
+
+    name: str
+        Remote file name
+
+    file_obj
+        File-like object to upload.
+
+    Returns
+    -------
+    None
+
+    """
+    client.putfo(fl=file_obj, remotepath=name)
+
+
+def lambda_handler(event, context):
+    """Sample pure Lambda function
+
+    Parameters
+    ----------
+    event: dict, required
+        The EventBridge event payload containing the following required keys:
+        - ssm_secret_prefix: str
+            The prefix for fetching secrets from AWS Systems Manager Parameter Store.
+        - period_count: int
+            The number of months to retrieve balances for. Must be greater than 0.
+
+    context: object, required
+        Lambda Context runtime methods and attributes.
+        See AWS Lambda documentation for details:
+        https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html
+
+    Returns
+    -------
+    None
+    """
+
+    # number of months to get balances for
+    period_count = get_period_count(event)
+    logging.info(f"Getting balances for past {period_count} months")
+
+    # get secrets from SSM and authenticate
+    if "ssm_secret_prefix" not in event:
+        raise ValueError("'ssm_secret_prefix' not provided")
+    ssm_prefix = event["ssm_secret_prefix"]
+    auth = get_ssm_params(ssm_prefix)
+
+    logging.info(f"Logging in to SFTP server")
+    client = get_sftp_client(auth)
+
+    # Start with today, and go back N months
+    try:
+        period = date.today()
+        for _ in range(period_count):
+            when = period.isoformat()
+            name, file_obj = get_balances_csv(when)
+            logging.info(f"Uploading {name} to SFTP server")
+            put_sftp_file(client, name, file_obj)
```

Comment:
how about logging failures?  also do you think we need to handle failures or just report them?

Reply:
oh, great catch, we should definitely add more logging around upload errors, this is currently just relying on the lambda environment logging a generic exception and exiting.
but I'm not sure there's much handling we can do other than logging. and I would prefer for the exceptions to be fatal because if one upload fails, it's very unlikely that the next upload would succeed, and repeated failures might lock us out.